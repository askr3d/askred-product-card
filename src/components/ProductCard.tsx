import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';
import { InitialValues, onChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../types';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } =  ProductContext;


export type Props = {
    product: Product,
    // children?: ReactElement | ReactElement[]
    children: (args: ProductCardHandlers) => React.JSX.Element
    className?: string
    style?: CSSProperties
    onChange?: (args: onChangeArgs) => void
    value?: number
    initialValues?: InitialValues
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {
    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

  return (
    <Provider value={{ 
        counter,
        increaseBy,
        product,
        maxCount
    }}>
        <div 
            className={ `${ styles.productCard } ${ className }`}
            style={ style }
        >
            {
                children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,

                    increaseBy,
                    reset
                })
            }
        </div>
    </Provider>
  )
}
