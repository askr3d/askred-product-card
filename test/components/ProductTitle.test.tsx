import React from "react";
import ProductCard, { ProductTitle } from '../../src/components';
import { render } from "@testing-library/react";
import { product1 } from "../data/products";

describe('ProductTitle', () => { 
    test('debe de mostrar el componente correctamente con el titulo personalizado', () => {
        const { asFragment } = render(
            <ProductTitle title='Custom Title' className="custom-class" />
        )
        
        expect( asFragment() ).toMatchSnapshot();
    });

    test('debe de mostrar el componente con el nombre del producto', () => {
        const { asFragment } = render(
            <ProductCard product={product1}>
                {
                    () => (
                        <ProductTitle />
                    )
                }
            </ProductCard>
        )
        
        expect( asFragment() ).toMatchSnapshot();
    });
});