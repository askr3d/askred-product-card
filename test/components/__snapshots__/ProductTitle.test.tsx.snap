// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe de mostrar el componente con el nombre del producto 1`] = `
<DocumentFragment>
  <div
    class="productCard undefined"
  >
    <span
      class=" productDescription undefined "
    >
      Coffee Mug - Sin Imagen
    </span>
  </div>
</DocumentFragment>
`;

exports[`ProductTitle debe de mostrar el componente correctamente con el titulo personalizado 1`] = `
<DocumentFragment>
  <span
    class=" productDescription custom-class "
  >
    Custom Title
  </span>
</DocumentFragment>
`;
